lib/puppet/util/character_encoding.rb

```diff
@@ -62,13 +63,27 @@ def convert_to_utf_8!(string)
           # (it is valid in its current encoding), we can guess this string was
           # not originally unicode. Transcode it to UTF-8. For strings with
           # original encodings like SHIFT_JIS, this should be the final result.
+          # If the string comes to us as BINARY encoded, we don't know what it
+          # started as. However, to encode! we need a starting place, and our
+          # best guess (only guess) is whatever the system currently is
+          # (default_external). So set external_encoding to default_external
+          # before we try to transcode to UTF-8. There's one exception, which is
+          # if external encoding is currently UTF-8. We don't bother
+          # force_encoding to UTF-8 first because: a) once we do, ruby won't
+          # transcode it for us to UTF-8 anymore and b) we've already previously
+          # determined that this string is not valid UTF-8 bytes so force
+          # encoding to UTF-8 isn't going to help us
+          string.force_encoding(Encoding.default_external) if string.encoding == Encoding::BINARY && Encoding.default_external != Encoding::UTF_8
           return string.encode!(Encoding::UTF_8)
         else
           # If the string is neither valid UTF-8 as-is nor valid in its current
           # encoding, fail. It requires user remediation.
           raise EncodingError
         end
       rescue EncodingError => detail
+        # Ensure the string retains it original external encoding since
+        # we've failed
+        string.force_encoding(original_encoding)
```

Comment:
@joshcooper I'm sorry I'm not connecting the dots. Are you saying this line would end up modifying the string from its original state? My hope was that it was actually doing the opposite - restoring its original state after it was modified on line 76 in anticipation of calling `encode!`. I could modify line 86 to only do `force_encoding` if the same criteria are satisfied as we demand on line 76, ie only if the string's original encoding was BINARY and default external is not UTF-8? Is that what you mean?